src/som/interpreter/actors/EventualSendNode.java

```diff
@@ -183,24 +196,72 @@ protected void sendDirectMessage(final Object[] args, final Actor owner,
       assert !(args[0] instanceof SFarReference) : "This should not happen for this specialization, but it is handled in determineTargetAndWrapArguments(.)";
       assert !(args[0] instanceof SPromise) : "Should not happen either, but just to be sure";
 
+      SteppingType stepping = validateResolverForStepping(resolver);
+
       DirectMessage msg = new DirectMessage(
-          EventualMessage.getCurrentExecutingMessageId(), target, selector, args,
+          EventualMessage.getCurrentExecutingMessage(), EventualMessage.getCurrentExecutingMessageId(), target, selector, args,
           owner, resolver, onReceive,
-          messageReceiverBreakpoint.executeCheckIsSetAndEnabled(), promiseResolverBreakpoint.executeCheckIsSetAndEnabled());
+          messageReceiverBreakpoint.executeCheckIsSetAndEnabled(), promiseResolverBreakpoint.executeCheckIsSetAndEnabled(), stepping);
+
       target.send(msg, actorPool);
     }
 
     protected void sendPromiseMessage(final Object[] args, final SPromise rcvr,
         final SResolver resolver, final RegisterWhenResolved registerNode) {
       assert rcvr.getOwner() == EventualMessage.getActorCurrentMessageIsExecutionOn() : "think this should be true because the promise is an Object and owned by this specific actor";
 
+      SteppingType stepping = validateResolverForStepping(resolver);
+
       PromiseSendMessage msg = new PromiseSendMessage(
-          EventualMessage.getCurrentExecutingMessageId(), selector, args,
+          EventualMessage.getCurrentExecutingMessage(), EventualMessage.getCurrentExecutingMessageId(),
+          selector, args,
           rcvr.getOwner(), resolver, onReceive,
-          messageReceiverBreakpoint.executeCheckIsSetAndEnabled(), promiseResolverBreakpoint.executeCheckIsSetAndEnabled());
+          messageReceiverBreakpoint.executeCheckIsSetAndEnabled(), promiseResolverBreakpoint.executeCheckIsSetAndEnabled(), stepping);
+
       registerNode.register(rcvr, msg, rcvr.getOwner());
     }
 
+    /**
+     * Validates if the message has null resolver
+     * when a step-to-promise-resolution has been made, in that case
+     * none stepping strategy is returned.
+     */
+    private SteppingType validateResolverForStepping(final SResolver resolver) {
+      if (!VmSettings.TRUFFLE_DEBUGGER_ENABLED) {
+       return null;
+      }
+
+      SteppingStrategy strategy = ActivityThread.steppingStrategy();
+      SteppingType stepping = null;
+      boolean stop = false;
+
+      if (strategy == null) {
+        return null;
+      }
+
+      if (strategy instanceof ToMessageReceiver) {
```

Comment:
it is really unclear to me why you first do a `instanceof` check and then use a polymorphic method that could do everything you need.